### ISSUE TITLE
killjava.sh not invoked on java.lang.OutOfMemoryError: unable to create new native thread

### ISSUE THREAD

**Issue by youngm** [cloudfoundry/java-buildpack#161]
[This stack overflow topic](http://stackoverflow.com/questions/23373117/xxonoutofmemoryerror-doesnt-work-for-java-lang-outofmemoryerror-unable-to-cr) theorizes that the `OnOutOfMemoryError` process cannot be created because there are no more threads available to the OS in this situation.

Perhaps we could handle this case for simple `.war` files by implementing the OOM script in a tomcat valve then invoking `System.exit()` in the case of an OOME so that the process exits and the instance restarts?

Spring Boot apps and such would be on their own in this case but perhaps Spring Boot could provide something general here.

**Comment by youngm** [comment]
Actually I suppose as a valve this may not be very useful.  In the case of creating a thread it is probable that the call trace may be in another thread where the Valve is not in the call chain.  Oh well.  I guess we can log this issue away as interesting data and close it I suppose.

I don't have any ideas of how to mitigate this issue.

**Comment by youngm** [comment]
Perhaps a better solution that may work in more cases than just tomcat Is to create a java agent that sets a default UncaughtExceptionHandler [0] that can find uncaught OOMEs and System.exit().

I recently had another application get an OOME not get handled by OnOutOfMemoryError.  This error message looked like this:
```
OpenJDK 64-Bit Server VM warning: Exception java.lang.OutOfMemoryError occurred dispatching signal SIGTERM to handler- the VM may need to be forcibly terminated
```

Perhaps a default UncaughtExeptionHandler would handle this type of OOME as well?

[0] http://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html#setDefaultUncaughtExceptionHandler-java.lang.Thread.UncaughtExceptionHandler-

**Comment by cgfrost** [comment]
FYI, Issue created https://www.pivotaltracker.com/story/show/90196742

**Comment by nebhale** [comment]
@youngm We've not tried anything in regards to this (and I'm not sure how definitive we could be anyway), but it seems that the OOM situation would preclude the creation of the `Runtime.exec()` infrastructure to execute the script (or equivalent).  Specifically, we'd need to create buffers to capture `stdout` and `stderr` which require more memory.

What's your read on that?

**Comment by youngm** [comment]
@nebhale I'm not exactly sure that you're getting at with the `stdout` and `stderr` buffers.  So, let me try to explain the issue another way to be sure we're on the same page.  My thoughts evolved a bit since I initially created the issue and I want to be sure they are all summarized correctly.

It seems in the JVM there are many situations that can prompt an OOME.  In the buildpack today we use the `-XX:OnOutOfMemoryError` switch to kill the JVM so that CF will restart it which should work great it most all OOME situations.  However, it appears that the one OOME that cannot be handled by `-XX:OnOutOfMemoryError` is the case of a `unable to create new native thread`.  It appears this OOME exception happens when the system runs out of threads/processes.  The problem is that because the system is out of processes a process to execute the `-XX:OnOutOfMemoryError` cannot be spawned.  Leaving the app running in a broken state with no way for CF to know it needs to be restarted.

I believe we could catch and handle this type of OOME by setting a  `Thread.setDefaultUncaughtExceptionHandler()`.  In theory an UncauchtExceptionHandler could do most everything (and perhaps more) than the `killjava.sh`  script is doing and it could also in theory not require the creation of a new thread allowing it to handle the `unable to create new native thread` case.

Now the question is how do we implement such a handler that will not itself fail to execute because it runs out of memory while executing.

This is what I'm thinking as a possible solution:
* Keep the current `-XX:OnOutOfMemoryError` as a final last try catch all.
* Create a JavaAgent that sets and implements a `Thread.setDefaultUncaughtExceptionHandler()`
* Implement the `UncaughtExceptionHandler` such that it makes a decent effort to not consume too much memory and definitely creates no Threads.  I don't think we need to go too crazy on the not consume memory part because if the OOME were caused by too many threads then the JVM should have plenty of heap to complete.  If the OOME were caused by running out of heap then the UncaughtExceptionHandler may run out of memory itself while trying to gather diagnostics.  If it does then no big deal `killjava.sh` should still execute gathering diagnostics and terminating the jvm.

Here is some potential pseudo code for the `UncaughtExceptionHandler`

```
String ulimitOutput;
OOMEExceptionHandler() {
ulimitOutput = System.exec("ulimit -a")
}
void uncaughtException(Thread t, Throwable e) {
  if(!(e instanceof OOME)) {
    return;
  }
  //Build Diagnostics output
  System.err.println("ulimit data:"+ulimitOutput);
  System.err.println("Disk Space:"+new File().getFreeSpace());
  //etc
  //Unfortunately I don't see a clean way to get process status without creating a process.
  System.exit(1);
}
```

The keys here I think are:
* NOT putting System.exit() in a try finally so that `-XX:OnOutOfMemoryError` will still get executed if the handler causes another heap based OOME.
* Gather what data you can (like ulimit -a) prior to OOME.

Thoughts?

**Comment by nebhale** [comment]
So, the point I was trying to make (badly, clearly) is well illustrated by your example.  Specifically `System.exec("ulimit -a")` needs to consume some memory for the output of that command.  I'm not sure what the implementation of `System.exec()` is under the hood, but if it is based on `Runtime.exec()`, then you'll [need two extra threads](http://stackoverflow.com/a/9048519/1384449) to pop data off of the `stdout` and `stderr` sockets.

I'm not sure a thought-experiment is going to cut it on this though; we probably need to actually implement and test the _actual_ behavior...

**Comment by youngm** [comment]
System.exec was a typo I meant Runtime.exec().  And yes.  I think for the `UncaughtExceptionHandler` to work I think we'd need to limit it to data that can be gathered without executing a Runtime.exec() unfortunately since the main value this handler would be to handle the case where the system is out or processes.

Unless you feel good about parsing the OOME message to determine if it is a thread constrained error or not.  If not we could just rethrow the exception and let `-XX:OnOutOfMemoryError` handle it.

Yes, I think some experiments are in order.

**Comment by youngm** [comment]
Perhaps something like this could help also: https://github.com/airlift/jvmkill

**Comment by rafaelri** [comment]
@youngm I tried https://github.com/airlift/jvmkill and it worked just fine. I've created a fork of the java buildpack with a Component for configuring the agent and it is available at https://github.com/rafaelri/java-buildpack . I noticed when trying to create a pull request that I need to sign a Contributor License Agreement before creating pull requests. Please have a look at my attempt and let me know if something needs to be changed and also how to sign the agreeement.
I'd also like to point something in regard to using the agent instead of an UncaughtExceptionHandler. With the agent we don't risk that any framework or code changes the handler rendering our mechanism useless. On the other hand, the agent is native code and we have to pay attention between JVM releases, although after the arrival of the JVMTI interface this became quite stable. I also took some time to build a rather simple web application to exercise this issue, it is available at https://github.com/rafaelri/webapp-oom
kind regards,
Rafael

**Comment by rafaelri** [comment]
@youngm forgot to mention on the previous comment that I did some tests using only the agent without the kill_java script (I've deleted a previous fork with only the agent) and it worked just fine but after seeing cloudfoundry/java-buildpack#156 I thought that perharps we could have some environment variable in which we could choose between shell script, agent or none. I'll place my comments regarding having none on the appropriate issue (cloudfoundry/java-buildpack#156). 
I think that by having the two possibilities we can still let the user opt for the shell script in case there are any incompatibilities with a new JVM release.

**Comment by youngm** [comment]
@rafaelri This looks great!  I hope you can sign the CLA and get a PR put together.

**Comment by rafaelri** [comment]
@youngm sorry for the long delay, I had a few tough weeks at work since I'll be on vacation next week then I'll definitely have time to take a look at this. In the meantime I found the CLA and I'll sign and mail it. Also, I'd like to have a careful look in the changes mentioned by Christian in the heap parameters of the VM. I'll also rewrite my fork incorporating the agent in the OpenJdkLike component as I feel it fits better there (even more when we consider the possibility of switching between the script, agent or none).

**Comment by youngm** [comment]
@rafaelri It would be nice to get some feedback from @cgfrost or @nebhale about this approach since they'd be the ones to actually merge such a PR.

**Comment by nebhale** [comment]
Yup, we'd be interested in seeing the PR.  @rafaelri you're correct that it should go in as a component of the `OpenJdkLike` rather than as a separate component.  This does feel like a much better solution than we're currently dealing with.

**Comment by rafaelri** [comment]
@nebhale I've just reimplemented as we agreed. Please have a look and check if it is fine. I've an application deployed with the changed buildpack: http://rafaelri.cfapps.io/oot (triggers the error due to a huge amount of threads), http://rafaelri.cfapps.io/oom (heap exhausted) and http://rafaelri.cfapps.io/hello (is a hello world to check proper application functioning after restart). I have to point that it is still missing the tests (I know they should have come first but as I am still getting used to the code I've postponed them). BTW I've printed the CLA and will send it soon.

**Comment by rafaelri** [comment]
@nebhale I guess I missed something. Some tests are failing and the error message complains about something related to the index.yml cached file for the agent (that btw I've hosted in a github repo).

**Comment by glyn** [comment]
Hi @rafaelri. Did you send in the CLA?

**Comment by rafaelri** [comment]
Glyn I sent it in digital form I must admit that I haven't mailed it yet.
I can mail it this week.
Em 29/02/2016 09:15, "Glyn Normington" <notifications@github.com> escreveu:

**Comment by glyn** [comment]
@rafaelri someone sent in an individual CLA by email recently and got back an email response saying "Your GitHub ID (...) has been added to our whitelist.". Did you get a similar reply? Was it a corporate CLA?

@nebhale: I don't know how to check the whitelist. Is there a bot which polls the whitelist and updates github PRs?

**Comment by rafaelri** [comment]
@glyn searched here and couldn't find such e-mail. I also ended up signing
two CLAs (both individual ones) cause I have signed the first one
immediately before the change from Pivotal to Cloud Foundry; Chip Childers
contacted me providing the newest one.

2016-02-29 9:42 GMT-03:00 Glyn Normington <notifications@github.com>:

**Comment by kelapure** [comment]
@glyn  Any updates here ?

**Comment by nebhale** [comment]
@glyn and @cgfrost are actively working on this.

**Comment by kelapure** [comment]
@nebhale  Thanks for the update

**Comment by glyn** [comment]
@kelapure @cgfrost and I work on the Java buildpack for the majority of every Monday, although last Monday was a public holiday. We should be back on it next Monday...

**Comment by youngm** [comment]
FWIW, I've been doing some testing with Diego/Garden.  Apparently Diego/Garden doesn't currently set a limit on processes for a container.  Which makes this issue a little less important.  Though I believe the CAPI team is planning to eventually make process limits in containers configurable which may make this issue important again.  Though I don't have a story link for that currently.

**Comment by glyn** [comment]
@youngm Thanks. Diego/Garden may refer to garden-linux. garden-runc does set a limit on the number of processes for a container if the user sets ContainerSpec.Limits.Pids, which I understand is the case in the most recent release of CF. (This will mitigate fork bombs.)

Anyway, the JVMTI agent is now complete and waiting to be integrated into the Java buildpack. I understand @nebhale intends this integration to occur before the next significant release of the buildpack.

**Comment by youngm** [comment]
Thanks @glyn !

**Comment by nebhale** [comment]
Resolved via 03fccb8d5671d3988f98b9226a65314dc5afa771
